ts2phc.8
.TH TS2PHC 8 "December 2019" "linuxptp"
.SH NAME
ts2phc - Synchronizes one or more PTP Hardware Clocks using external time stamps.

.SH SYNOPSIS
.B ts2phc
[
.B \-hmqv
] [
.BI \-c " device|name"
] [
.BI \-f " config"
] [
.BI \-l " print-level"
] [
.BI \-s " device|name"
] [
.I long-options
]
.I .\|.\|.

.SH DESCRIPTION
.B ts2phc
synchronizes PTP Hardware Clocks (PHC) to external time stamp signals.
A single source may be used to distribute time to one or more PHC devices.

.SH OPTIONS
.TP
.BI \-c " device|name"
Specifies a PHC slave clock to be synchronized.
The clock may be identified by its character device (like /dev/ptp0)
or its associated network interface (like eth0).
This option may be given multiple times.
.TP
.BI \-f " config"
Read configuration from the specified file.
No configuration file is read by default.
.TP
.BI \-h
Displays the command line help summary.
.TP
.BI \-l " print-level"
Sets the maximum syslog level of messages which should be printed or
sent to the system logger. The default is 6 (LOG_INFO).
.TP
.B \-m
Prints log messages to the standard output.
.TP
.B \-q
Prevents sending log messages to the system logger.
.TP
.BI \-s " device|name"
Specifies the source of the PPS signal.
Use the key word "generic" for an external 1-PPS without ToD information.
When using a master PHC, the clock may be identified by its character
device (like /dev/ptp0) or its associated network interface (like
eth0).
.TP
.B \-v
Prints the software version and exits.

.SH LONG OPTIONS

Each and every configuration file option (see below) may also appear
as a "long" style command line argument.  For example, the use_syslog
option may be set using either of these two forms.

.RS
\f(CW\-\-use_syslog 1   \-\-use_syslog=1\fP
.RE

Option values given on the command line override values in the global
section of the configuration file.

.SH CONFIGURATION FILE

The configuration file is divided into sections. Each section starts with a
line containing its name enclosed in brackets and it follows with settings.
Each setting is placed on a separate line, it contains the name of the
option and the value separated by whitespace characters. Empty lines and lines
starting with # are ignored.

There are two different section types.

.TP
.B 1.
The global section (indicated as
.BR [global] )
sets the program options and default slave clock options. Other
sections are clock specific sections, and they override the default
options.
.TP
.B 2.
Slave clock sections give the name of the configured slave (e.g.
.BR [eth0] ).
Slave clocks specified in the configuration file need not be specified
with the
.B \-c
command line option.

.SH GLOBAL OPTIONS

.TP
.B first_step_threshold
The maximum offset, specified in seconds, that the servo will correct
by changing the clock frequency instead of stepping the clock. This is
only applied on the first update. When set to 0.0, the servo will not
step the clock on start.
The default is 0.00002 (20 microseconds).
.TP
.B free_running
When set to 1, no the slave clock will be adjusted.
This option can be useful in test scenarios, for example to determine
how well synchronized a group of local clocks are to each other.
The default is 0 (adjust the slave clocks).
.TP
.B logging_level
The maximum logging level of messages which should be printed.
The default is 6 (LOG_INFO).
.TP
.B max_frequency
The maximum allowed frequency adjustment of the clock in parts per
billion.  This is an additional limit to the maximum allowed by the
hardware. When set to 0, the hardware limit will be used.
The default is 900000000 (90%).
.TP
.B message_tag
The tag which is added to all messages printed to the standard output
or system log.  The default is an empty string (which cannot be set in
the configuration file as the option requires an argument).
.TP
.B step_threshold
The maximum offset, specified in seconds, that the servo will correct
by changing the clock frequency instead of stepping the clock. When
set to 0.0, the servo will never step the clock except on start.
The default is 0.0.
.TP
.B ts2phc.pulsewidth
The expected pulse width of the external PPS signal in nanoseconds.
When 'ts2phc.extts_polarity' is "both", the given pulse width is used
to detect and discard the time stamp of the unwanted edge.
The supported range is 1000000 to 990000000 nanoseconds.
The default is 500000000 nanoseconds.
.TP
.B use_syslog
Print messages to the system log if enabled.  The default is 1 (enabled).
.TP
.B verbose
Print messages to the standard output if enabled.  The default is 0 (disabled).

.SH SLAVE CLOCK OPTIONS

.TP
.B ts2phc.channel
The external time stamping or periodic output channel to be used.
Some PHC devices feature programmable pins and one or more time
stamping channels.  This option allows selecting a particular channel
to be used.  When using a PHC device as the PPS source, this option
selects the periodic output channel.
The default is channel 0.
.TP
.B ts2phc.extts_correction
The value, in nanoseconds, to be added to each PPS time stamp.
The default is 0 (no correction).
.TP
.B ts2phc.extts_polarity
The polarity of the external PPS signal, either "rising" or "falling".
Some PHC devices always time stamp both edges.  Setting this option to
"both" will allow the ts2phc program to work with such devices by
detecting and ignoring the unwanted edge.  In this case be sure to
set 'ts2phc.pulsewidth' to the correct value.
The default is "rising".
.TP
.B ts2phc.master
Setting this option to 1 configures the given PHC device as the source
of the PPS signal.
The default is 0 for the slave role.
.TP
.B ts2phc.pin_index
The pin index to be used.
Some PHC devices feature programmable pins, and this option allows
configuration of a particular pin for the external time stamping or
periodic output function.
The default is pin index 0.

.SH WARNING

Be cautious when sharing the same configuration file between ptp4l,
phc2sys, and ts2phc.  Keep in mind that values specified in the
configuration file take precedence over the default values.  If an
option which is common to the other programs is set in the
configuration file, then the value will be applied to all the programs
using the file, and this might not be what is expected.

It is recommended to use separate configuration files for ptp4l,
phc2sys, and ts2phc in order to avoid any unexpected behavior.

.SH SEE ALSO
.BR phc2sys (8)
.BR ptp4l (8)


ts2phc.c
/**
 * @file ts2phc.c
 * @brief Utility program to synchronize the PHC clock to external events
 * @note Copyright (C) 2013 Balint Ferencz <fernya@sch.bme.hu>
 * @note Based on the phc2sys utility
 * @note Copyright (C) 2012 Richard Cochran <richardcochran@gmail.com>
 * @note SPDX-License-Identifier: GPL-2.0+
 */
#include <stdlib.h>

#include "config.h"
#include "print.h"
#include "ts2phc_master.h"
#include "ts2phc_slave.h"
#include "version.h"

static void ts2phc_cleanup(struct config *cfg, struct ts2phc_master *master)
{
	ts2phc_slave_cleanup();
	if (master) {
		ts2phc_master_destroy(master);
	}
	if (cfg) {
		config_destroy(cfg);
	}
}

static void usage(char *progname)
{
	fprintf(stderr,
		"\n"
		"usage: %s [options]\n\n"
		" -c [dev|name]  phc slave clock (like /dev/ptp0 or eth0)\n"
		"                (may be specified multiple times)\n"
		" -f [file]      read configuration from 'file'\n"
		" -h             prints this message and exits\n"
		" -l [num]       set the logging level to 'num'\n"
		" -m             print messages to stdout\n"
		" -q             do not print messages to the syslog\n"
		" -s [dev|name]  source of the PPS signal\n"
		"                may take any of the following forms:\n"
		"                    generic   - an external 1-PPS without ToD information\n"
		"                    /dev/ptp0 - a local PTP Hardware Clock (PHC)\n"
		"                    eth0      - a local PTP Hardware Clock (PHC)\n"
		"                    nmea      - a gps device connected by serial port or network\n"
		" -v             prints the software version and exits\n"
		"\n",
		progname);
}

int main(int argc, char *argv[])
{
	char *config = NULL, *pps_source = NULL, *progname;
	int c, err = 0, have_slave = 0, index, print_level;
	struct ts2phc_master *master = NULL;
	enum ts2phc_master_type pps_type;
	struct config *cfg = NULL;
	struct interface *iface;
	struct option *opts;

	handle_term_signals();

	cfg = config_create();
	if (!cfg) {
		ts2phc_cleanup(cfg, master);
		return -1;
	}

	opts = config_long_options(cfg);

	/* Process the command line arguments. */
	progname = strrchr(argv[0], '/');
	progname = progname ? 1 + progname : argv[0];
	while (EOF != (c = getopt_long(argc, argv, "c:f:hi:l:mqs:v", opts, &index))) {
		switch (c) {
		case 0:
			if (config_parse_option(cfg, opts[index].name, optarg)) {
				ts2phc_cleanup(cfg, master);
				return -1;
			}
			break;
		case 'c':
			if (ts2phc_slave_add(cfg, optarg)) {
				fprintf(stderr, "failed to add slave\n");
				ts2phc_cleanup(cfg, master);
				return -1;
			}
			have_slave = 1;
			break;
		case 'f':
			config = optarg;
			break;
		case 'l':
			if (get_arg_val_i(c, optarg, &print_level,
					  PRINT_LEVEL_MIN, PRINT_LEVEL_MAX)) {
				ts2phc_cleanup(cfg, master);
				return -1;
			}
			config_set_int(cfg, "logging_level", print_level);
			print_set_level(print_level);
			break;
		case 'm':
			config_set_int(cfg, "verbose", 1);
			print_set_verbose(1);
			break;
		case 'q':
			config_set_int(cfg, "use_syslog", 0);
			print_set_syslog(0);
			break;
		case 's':
			if (pps_source) {
				fprintf(stderr, "too many PPS sources\n");
				ts2phc_cleanup(cfg, master);
				return -1;
			}
			pps_source = optarg;
			break;
		case 'v':
			ts2phc_cleanup(cfg, master);
			version_show(stdout);
			return 0;
		case 'h':
			ts2phc_cleanup(cfg, master);
			usage(progname);
			return -1;
		case '?':
		default:
			ts2phc_cleanup(cfg, master);
			usage(progname);
			return -1;
		}
	}
	if (config && (c = config_read(config, cfg))) {
		fprintf(stderr, "failed to read config\n");
		ts2phc_cleanup(cfg, master);
		return -1;
	}
	print_set_progname(progname);
	print_set_tag(config_get_string(cfg, NULL, "message_tag"));
	print_set_verbose(config_get_int(cfg, NULL, "verbose"));
	print_set_syslog(config_get_int(cfg, NULL, "use_syslog"));
	print_set_level(config_get_int(cfg, NULL, "logging_level"));

	STAILQ_FOREACH(iface, &cfg->interfaces, list) {
		if (1 == config_get_int(cfg, iface->name, "ts2phc.master")) {
			if (pps_source) {
				fprintf(stderr, "too many PPS sources\n");
				ts2phc_cleanup(cfg, master);
				return -1;
			}
			pps_source = iface->name;
		} else {
			if (ts2phc_slave_add(cfg, iface->name)) {
				fprintf(stderr, "failed to add slave\n");
				ts2phc_cleanup(cfg, master);
				return -1;
			}
			have_slave = 1;
		}
	}
	if (!have_slave) {
		fprintf(stderr, "no slave clocks specified\n");
		ts2phc_cleanup(cfg, master);
		usage(progname);
		return -1;
	}
	if (!pps_source) {
		fprintf(stderr, "no PPS source specified\n");
		ts2phc_cleanup(cfg, master);
		usage(progname);
		return -1;
	}
	if (ts2phc_slave_arm()) {
		fprintf(stderr, "failed to arm slaves\n");
		ts2phc_cleanup(cfg, master);
		return -1;
	}

	if (!strcasecmp(pps_source, "generic")) {
		pps_type = TS2PHC_MASTER_GENERIC;
	} else if (!strcasecmp(pps_source, "nmea")) {
		pps_type = TS2PHC_MASTER_NMEA;
	} else {
		pps_type = TS2PHC_MASTER_PHC;
	}
	master = ts2phc_master_create(cfg, pps_source, pps_type);
	if (!master) {
		fprintf(stderr, "failed to create master\n");
		ts2phc_cleanup(cfg, master);
		return -1;
	}

	while (is_running()) {
		err = ts2phc_slave_poll(master);
		if (err) {
			pr_err("poll failed");
			break;
		}
	}

	ts2phc_cleanup(cfg, master);
	return err;
}
